tests/utils.py

```diff
@@ -414,3 +415,129 @@ def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> None:
                                     f" args {args} and kwargs {kwargs}")
 
     return wrapper
+
+
+async def completions_with_server_args(
+    prompts: List[str],
+    model_name: str,
+    server_cli_args: List[str],
+    num_logprobs: Optional[int],
+) -> Completion:
+    '''Construct a remote OpenAI server, obtain an async client to the
+    server & invoke the completions API to obtain completions.
+
+    Args:
+      prompts: test prompts
+      model_name: model to spin up on the vLLM server
+      server_cli_args: CLI args for starting the server
+
+    Returns:
+      OpenAI Completion instance
+    '''
+
+    outputs = None
+    with RemoteOpenAIServer(model_name, server_cli_args) as server:
+        client = server.get_async_client()
+        outputs = await client.completions.create(model=model_name,
+                                                  prompt=prompts,
+                                                  temperature=0,
+                                                  stream=False,
+                                                  max_tokens=5,
+                                                  logprobs=num_logprobs)
+    assert outputs is not None
+
+    return outputs
+
+
+def get_client_text_generations(completions: Completion):
+    '''Extract generated tokens from the output of a
+    request made to an Open-AI-protocol completions endpoint.
+    '''
+    return [x.text for x in completions.choices]
+
+
+'''Logprobs values are extracted as
+List[Optional[List[Dict[str,float]]]], i.e.:
+
+For each :class:`SequenceGroup`...
+  ...if the completions API was invoked with a non-`None` `logprobs` argument:
+      ...for each token offset in a sequence...
+          ...store a mapping from str(token) -> logprob
+  ...else, if the completions API was invoked with `logprobs=None`:
+      ...store None
+'''
+LogprobType = List[Optional[List[Dict[str, float]]]]
+
+
+def get_client_logprob_generations(completions: Completion) -> LogprobType:
+    '''Operates on the output of a request made to an Open-AI-protocol
+    completions endpoint; obtains top-rank logprobs for each token in
+    each :class:`SequenceGroup`
+    '''
+    return [(None if x.logprobs is None else x.logprobs.top_logprobs)
+            for x in completions.choices]
+
+
+def assert_all_close_logprobs(
```

Comment:
We have `check_logprobs_close` in `test/models/utils.py` too. Could we merge these 2 together?

Reply:
I appreciate what you are getting at with this! However, it is worth noting that `check_logprobs_close()` has a significant difference in its logic & purpose.

`check_logprobs_close()` is designed for a scenario where you are testing for failures in model design compared to a HuggingFace baseline. It actually allows for significant differences in logprob values, because there can be differences in numerics & order-of-operations between vLLM and HuggingFace which lead to significant rounding errors. It does not test for scenarios where the output tokens are correct but the logprobs are wrong (which matters, given that my PR is adding logprobs support.) `check_logprobs_close()` *assumes* that if the tokens match, the logprobs match; therefore `check_logprobs_close()` *only* examines logprobs at a given sequence position if the tokens at that position don't match. Furthermore, when examining logprobs, `check_logprobs_close()` does not directly compare logprobs to ensure they are nearly identical; it only requires that the token in sequence A is in the top-K logprobs of sequence B, and visa versa. You can see this below:

```
def check_logprobs_close(
    *,
    outputs_0_lst: Sequence[TokensTextLogprobs],
    outputs_1_lst: Sequence[TokensTextLogprobs],
    name_0: str,
    name_1: str,
    num_outputs_0_skip_tokens: int = 0,
    warn_on_mismatch: bool = True,
):
        ...

        # Loop through generated tokens.
        for idx, (output_id_0,
                  output_id_1) in enumerate(zip(output_ids_0, output_ids_1)):

            # If generated tokens don't match, then
            if output_id_0 != output_id_1:
                logprobs_elem_0 = logprobs_0[idx]
                logprobs_elem_1 = logprobs_1[idx]

                # Each predicted token must be in top N logprobs of the other
                fail_msg = (
                    f"Test{prompt_idx}:"
                    f"\nMatched tokens:\t{output_ids_0[:idx]}"
                    f"\n{name_0}:\t{output_str_0!r}\t{logprobs_elem_0}"
                    f"\n{name_1}:\t{output_str_1!r}\t{logprobs_elem_1}")

                assert logprobs_elem_0 is not None, fail_msg
                assert logprobs_elem_1 is not None, fail_msg
                assert output_id_0 in logprobs_elem_1, fail_msg
                assert output_id_1 in logprobs_elem_0, fail_msg
```

I would argue that `check_logprobs_close()` is actually a misnomer; a better name would be `tokens_match_or_tokens_are_mutually_in_top_logprobs()`

In contrast, the all_close function which I wrote strives to show that the multi-step top-K logprobs are identical to single-step top-K logprobs (i.e. matching with atol=1e-5 and rtol=1e-5) at all token offsets, with no regard for whether the tokens match. I think this functionality is too different from `check_logprobs_close()` to be reconciled.

Reply:
So in short, unless you have strong objections, I would prefer to keep the code the way it is.

Reply:
Sure ofc. Maybe you could even help rename that one to better illustrate what it does...

Reply:
Agreed, done.